packages/library-base/src/commonMain/kotlin/io/realm/internal/BaseResults.kt

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright 2020 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm.internal
+
+import io.realm.RealmObject
+import io.realm.RealmResults
+import io.realm.internal.interop.Link
+import io.realm.internal.interop.NativePointer
+import io.realm.internal.interop.RealmCoreException
+import io.realm.internal.interop.RealmInterop
+import kotlinx.coroutines.channels.ChannelResult
+import kotlinx.coroutines.channels.SendChannel
+import kotlinx.coroutines.flow.Flow
+import kotlin.reflect.KClass
+
+/**
+ * TODO : query
+ */
+// FIXME API-QUERY Final query design is tracked in https://github.com/realm/realm-kotlin/issues/84
+//  - Lazy API makes it harded to debug
+//  - Postponing execution to actually accessing the elements also prevents query parser errors to
+//    be raised. Maybe we can get an option to prevalidate queries in the C-API?
+internal abstract class BaseResults<E : Any> constructor(
+    protected val realm: RealmReference,
+    internal val nativePointer: NativePointer,
+    protected val clazz: KClass<E>,
+    protected val mediator: Mediator,
+    protected val mode: Mode = Mode.RESULTS
+) : AbstractList<E>(), RealmResults<E>, Freezable<BaseResults<E>>, Thawable<BaseResults<E>>, Observable<BaseResults<E>>, RealmStateHolder {
+
+    enum class Mode {
+        // FIXME Needed to make working with @LinkingObjects easier.
+        EMPTY, // RealmResults that is always empty.
+        RESULTS // RealmResults wrapping a Realm Core Results.
+    }
+
+    abstract fun instantiateResults(
+        realmReference: RealmReference,
+        nativePointer: NativePointer,
+        clazz: KClass<E>,
+        mediator: Mediator
+    ): BaseResults<E>
+
+    override val size: Int
+        get() = RealmInterop.realm_results_count(nativePointer).toInt()
+
+    @Suppress("SpreadOperator")
+    override fun query(query: String, vararg args: Any?): BaseResults<E> {
+        try {
+            return instantiateResults(
+                realm,
+                RealmInterop.realm_query_find_all(
```

Comment:
I suspect we don't want to call find_all here, but just do query_parse.

Reply:
This is the old "objects/query" API and should be removed.